content/news/2022-04-12-bevy-0.7/index.md

```diff
@@ -0,0 +1,383 @@
+
+In **Bevy 0.7**, we added the ability to automatically "upgrade" to using unbounded storage buffers for Clustered Forward Rendering on platforms that support them, enabling unlimited* point lights. There is an asterisk there because in practice this is limited by memory and hardware constraints. 
+
+## Light Clustering Features and Optimizations
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @robtfm, @dataphract, @cart</div>
+
+With the upper limit of 256 point lights removed, the only limit on lights is what the hardware can support and bottlenecks in our algorithms. To increase the number of lights, we made a number of optimizations to our clustering algorithms.
+
+* **Dynamic Light Clusters**: By default cluster x/y slices are now dynamically configured based on the lights in the scene, which can significantly increase performance in some scenes. The clustering behavior is now also user-configurable as FixedZ (the new default dynamic x/y behavior), custom fixed x/y/z slice values, single-cluster, and "no clustering".
+* **Light Frustum Change Detection**: We now use Bevy ECS's change detection feature to only recalculate the view frustum of lights that have changed.
+* **Cluster Assignment Optimizations**: The cluster assignment data access patterns and data structures received a variety of tweaks that improved performance.
+
+// TODO: Show many lights video here
+// TODO: Add iterative sphere refinement here if it is merged
+
+## Configurable Light Visibility 
+
+<div class="release-feature-authors">authors: @robtfm</div>
+
+Lights can now be turned on and off using Bevy's standard [`Visibility`] component:
+
+```rust
+commands.spawn(PointLightBundle {
+    visibility: Visibility {
+        is_visible: false,
+    },
+    ..default()
+});
+```
+
+[`Visibility`]: https://docs.rs/bevy/0.7.0/bevy/render/view/struct.Visibility.html
+
+## Auto-Labeled Systems For Nicer System Ordering
+
+<div class="release-feature-authors">authors: @cart, @aevyrie, @alice-i-cecile, @DJMcNab</div>
+
+Bevy uses "labels" to define ordering constraints between its ECS systems when they run in parallel. In previous versions of Bevy, the only option was to define custom labels:
+
+```rust
+#[derive(SystemLabel, Clone, Hash, Debug, PartialEq, Eq)]
+struct UpdateVelocity;
+
+app
+  .add_system(update_velocity.label(UpdateVelocity))
+  .add_system(movement.after(UpdateVelocity))
+```
+
+In **Bevy 0.7**, systems are now automatically labeled with a `SystemTypeIdLabel` tied to their type. This enables much more ergonomic and clear system ordering:
+
+```rust
+app
+  .add_system(update_velocity)
+  .add_system(movement.after(update_velocity))
+```
+
+The Bevy ECS labeling system is powerful and there are still legitimate use cases for custom labels (such as labeling multiple systems with the same label). But most common use cases can take advantage of the ergonomic auto-labeling functionality.
```

Comment:
Agreed, it feels backwards. It presents internal implementation details *first*, and only gets to why users should care later.

I would propose starting with the user-facing implications. Something like: "In bevy 0.7, you no longer have to come up with a custom label every time you need to order a system relative to another system; for simple cases you can just use the system's function, just like when you added the system!"

Then, if you really want to get into the implementation details (tbh I'm not sure such level of detail is even appropriate for a release blog post), you can elaborate, after the code example: "... and here is how this works. We have a `SystemTypeIdLabel` type ... ..."

In general, I think this sort of mindset should be applied everywhere. The news posts would read much better if they start by presenting the user-facing changes and motivations, and then maybe dive deeper into how it was done / implementation details (but not too much). Doing it the other way around feels backwards, as a reader. You are telling me a bunch of technical stuff, and I don't even know why I should care yet, until later.

Reply:
I agree that the inclusion of SystemTypeIdLabel is dubious the focus on "auto-labeling" as the mechanism of the feature distracts from the point.

But I do think the general "problem -> solution" format is a great way to illustrate progress because it tells a consistent story. I think covering the "this is what manual labeling looked like" first is important.

If you show the solution first, the user's first thoughts when they see it are no longer guaranteed to be "wow what a great improvement over the status quo". If they aren't familiar with the previous state of things, they might just think "yeah that tracks logically". 

Imagine this was a rendering feature that makes things much prettier. I think it generally makes sense to show the "older / uglier" thing first. That way we've primed the reader to discover things that are "better" in the new picture. The reverse order results in the reader looking for things that are "worse", which is a very different story. And not providing a contrasting image at all is an even bigger miss.

Definitely not saying either of you were actually suggesting removing the "before" picture here (or reordering it). Just providing rationale in case you were.

I'll try to reframe this to remove the focus on auto-labeling as the "after" picture and remove SystemTypeIdLabel. But I'll leave the introduction setting up the "before" picture (including discussion of labels) unless someone wants to push back on that.

Reply:
I do also agree with @inodentry that we should open with a something like "In bevy 0.7, you no longer have to come up with a custom label every time you need to order a system relative to another system; for simple cases you can just use the system's function, just like when you added the system!"

Reply:
Definitely not wanting to remove the before picture.
I'll be honest here, I did my first pass review quite quickly this morning, and completely skimmed over the before picture; that is, this comment is the first time I've even noticed it was there.

Reply:
Cool cool. I've changed how this is presented to make "auto labeling" conceptually an implementation detail, rather than the "feature" itself.